### ISSUE TITLE
RDF4J Server java.util.concurrent.ExecutionException

### ISSUE THREAD

**Issue by rknowles9** [eclipse/rdf4j#680]
I'm calling the RDF4J server from two different executions of a program that is updating some data on the server. The problem point is a clean-the-old before adding the new data. Both executions are at this code in my client at the same time. 

More discussion and stack traces from RDF4J versions 2.0.1 and 2.1.3 are shown at https://groups.google.com/forum/#!topic/rdf4j-users/qtqso7TFnWM

**Comment by jeenbroekstra** [comment]
As mentioned in the discussion: I have sofar been unable to reproduce the issue. OP will be looking into providing an MCVE.

**Comment by rknowles9** [comment]
This test is a Maven and Eclipse project. I ran it from the command line with
    => mvn clean install
    => mvn exec:java -Dexec.mainClass="org.rjk.test.rdf4jtest.Runner"

Runner starts two threads to run two executions in parallel. The main program uploads a large file to the server, then tries to clean up some possibly old data. The error on the server occurs during the clean up step. It seems dependent on how big the upload is. I tried reducing the size of the nt file; if it's much smaller there is no error; at this test size there's usually one thread successful and one fails; my production code, with a much bigger file, breaks in both threads. Once the exception is thrown on the server, i usually have to stop and restart it before I can test again or even get the workbench to access successfully.

Dick

[rdf4jtest.zip](https://github.com/eclipse/rdf4j/files/625887/rdf4jtest.zip)

**Comment by jeenbroekstra** [comment]
Thanks for the test case! I assume you closed it by mistake as the problem still exists :) I'll take a closer look as soon as possible.

**Comment by jeenbroekstra** [comment]
I can reproduce the problem. Investigating further.

**Comment by jeenbroekstra** [comment]
Ok, I still need to verify, but I *think* we may have unwittingly stumbled upon a critical error in how RDF4J server handles non-transactional requests. If I'm right, we'll have found (and hopefully fixed) a bug that may have been present since at least 2012...

**Comment by rknowles9** [comment]
Definitely closed by accident. First time on this site. I thought "Close and comment" meant "close this comment", not "close this issue and comment".
It looks like you're making progress on this. Thanks! We definitely need it since we expect to run that main process on perhaps hundreds of  ~140,000 statement models (not all at once, but definitely more than one at a time).
(About to press "Comment" and see if I get it right.)

**Comment by rknowles9** [comment]
How can I get hold of this for my work?- Wait for a release? Pull
(something) and build? Or ...?

**Comment by jeenbroekstra** [comment]
We will be releasing a patch release with this fix in it, probably middle of next week. However if you're able to take it for a test now, you can [grab a nightly build or build directly from source](http://rdf4j.org/download/#Source_code_and_nightly_builds) - would be good to get independent confirmation that the issue is indeed fixed.

**Comment by jeenbroekstra** [comment]
Btw if you want to test by just installing an updated RDF4J Server, you can just [download a snapshot build war file](https://oss.sonatype.org/content/repositories/snapshots/org/eclipse/rdf4j/rdf4j-http-server/2.2-SNAPSHOT/). Grab the most recent one war file there, download, rename (to `rdf4j-server.war`), then (re)deploy and you should be good to go. Not recommended in a production environment :)

**Comment by rknowles9** [comment]
I have grabbed the latest war file,
rdf4j-http-server-2.2-20161209.051730-24.war. So far running my test I'm
finding no errors reported in the test application or RDF4J server, but the
triples are not always being written. At first I was seeing both contexts
appearing when I ran the test from within eclipse, but when I ran from the
command line sometimes only one context and sometimes neither were written.
I added "conn.clear(context)" prior to writing to RDF4J, then I started to
get the intermittent results from both eclipse and command line. What I did
not do was update my test app to use the newer jars. I'm going to try to
pull and build and rerun the test with jars matching the war file.

**Comment by jeenbroekstra** [comment]
That sounds odd. If this problem persists, can you show me your updated code (if any) and/or what I should be querying for in the test db to verify that all expected data is present?

**Comment by rknowles9** [comment]
I've attached my updated test. I did not include the huge data file, since
I did not change that, but you'll still need it. I made changes to improve
logging and I added 'clearContext' to Uploader.

To check results: The same data is being loaded into two different
contexts. After a run I check Workbench. If everything loaded you'll see
two contexts named in 'Contexts' view, and 'Summary' will show :

    Repository Size
    Number of Statements    139998
    Number of Labeled Contexts    2

If you go to a context, then  pick any subject URI, the 'Explore' view that
comes up will show a set of quads in pairs, only different in context. For
example, exploring <
http://purl.rjk.test.org/sip/graphs/rjk-demo-opt/i/00000012> will show 4
triples, but each triple repeated in 2 quads, 1 for each context.

If only one context made it (which is what I'm usually seeing):

    Repository Size
    Number of Statements    69999
    Number of Labeled Contexts    1

and single sets of quads.

And, sometimes, nothing makes it:

    Repository Size
    Number of Statements    0
    Number of Labeled Contexts    0

These different results occur without error in my test code and without
error on RDF4J Server.

**Comment by jeenbroekstra** [comment]
@rknowles9 unfortunately GitHub is not clever enough to pick up e-mail attachments. You'll have to comment via the web UI and add the attachment that way...

**Comment by jeenbroekstra** [comment]
I've managed to reproduce some further related issues. I'm reopening this.

**Comment by jeenbroekstra** [comment]
Root cause is a concurrency/synchronization issue in SailSourceModel.

**Comment by jeenbroekstra** [comment]
Created a followup issue, #693, which is related not but not identical. Closing this, discussion can continue on the new issue. @rknowles9  thanks again for assisting with debugging this. I think I have it covered now. I'll ping you when I have a new snapshot build available, if you're available to take it for quick test spin again that'd be great.
